streambed-logged/src/lib.rs

```diff
@@ -32,6 +32,81 @@ const CONSUMER_QUEUE_SIZE: usize = 10;
 const PRODUCER_QUEUE_SIZE: usize = 10;
 const TOPIC_FILE_CONSUMER_POLL: Duration = Duration::from_secs(1);
 
+/// A compaction key is paired with an offset in a map. This map of
+/// compaction keys and their offsets express the minimum offset to
+/// retain as a result of having performed a scan of consumer records.
+#[derive(Clone, Hash, PartialEq, Eq)]
+pub struct CompactionKey {
+    pub topic: Topic,
+    pub partition: u32,
+    pub key: u64,
+}
+
+/// A compactor's role is to be fed consumer records and ultimately determine,
+/// for each record key, what the earliest offset is that may be retained. Upon
+/// the consumer completing, logged will then proceed to remove unwanted
+/// records from the commit log.
+pub struct Compactor<B> {
+    _init: B,
+    #[allow(clippy::type_complexity)]
+    _reducer: Box<dyn Fn(&mut B, &ConsumerRecord) -> bool>,
+    #[allow(clippy::type_complexity)]
+    _result: Box<dyn Fn(&B) -> HashMap<CompactionKey, u64>>,
+}
+
+impl<B> Compactor<B> {
+    /// Produce a new compactor with an initial state, a function for
+    /// reducing records to state representing the earliest records to retain,
+    /// and a function that maps that state to a map of keys to those offsets.
+    /// The reducer function receives a mutable state reference and a consumer
+    /// record, and returns a bool indicating whether the function has
+    /// reached its maximum number of distinct keys. If it has then
+    /// compaction may occur again once the current consumption of records
+    /// has finished. The goal is to avoid needing to process every type of
+    /// topic/partition/key i.e. a subset can be processed, and then another
+    /// subset etc.
+    /// The result function is responsible for mapping the state into
+    /// a map of keys and their minimum offsets.
+    pub fn new<F0, F1>(init: B, reducer: F0, result: F1) -> Self
+    where
+        F0: Fn(&mut B, &ConsumerRecord) -> bool + 'static,
+        F1: Fn(&B) -> HashMap<CompactionKey, u64> + 'static,
```

Comment:
I think the B should be consumed here.

Reply:
Also: again what is the u64?  It is an Offset right?  I would type alias it at least.  Same logic that has you define const's and not just insert the bare numbers in the code.   Just my 2c.